### PR TITLE
external: use multi-thread sort to speed up sorting

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: ref #45719

Problem Summary:
Need to speed up sorting, especially for small key-value.

### What is changed and how it works?
```
sorty.MaxGor = 8

thread quick sort 64.731125ms
quick sort 409.540667ms
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/writer_test.go [hunk 1]
```diff
@@ -414,3 +415,39 @@ func TestWriterMultiFileStat(t *testing.T) {
 	require.EqualValues(t, "key01", summary.Min)
 	require.EqualValues(t, "key24", summary.Max)
 }
+
+func TestWriterSort(t *testing.T) {
```
Comment (ywqzzy) [inline br/pkg/lightning/backend/external/writer_test.go:419]: Use bench？

Comment (wjhuang2016) [reply br/pkg/lightning/backend/external/writer_test.go:419]: I just want to make sure the data are sorted correctly.

File: br/pkg/lightning/backend/external/writer.go [hunk 2]
```diff
@@ -384,8 +384,16 @@ func (w *Writer) flushKVs(ctx context.Context, fromClose bool) (err error) {
 			zap.Any("rate", float64(savedBytes)/1024.0/1024.0/time.Since(ts).Seconds()))
 	}()
 
-	slices.SortFunc(w.writeBatch[:], func(i, j common.KvPair) int {
-		return bytes.Compare(i.Key, j.Key)
+	// TODO: set maxGoroutines according to the available CPU count.
+	sorty.MaxGor = 8
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/writer.go:388]: we already encoding/writing using configured concurrency, so each writer should restrict by it too.

**Comment by D3Hunter** [comment]
/hold

need discuss

File: br/pkg/lightning/backend/external/writer.go [hunk 3]
```diff
@@ -99,6 +100,7 @@ type WriterBuilder struct {
 	propKeysDist    uint64
 	onClose         OnCloseFunc
 	keyDupeEncoding bool
+	mu              *sync.Mutex
```
Comment (Benjamin2037) [inline br/pkg/lightning/backend/external/writer.go:103]: Please add comment for later resource control easy find out this mutex.

File: br/pkg/lightning/backend/local/local.go [hunk 4]
```diff
@@ -478,6 +478,7 @@ type Backend struct {
 	metrics      *metric.Common
 	writeLimiter StoreWriteLimiter
 	logger       log.Logger
+	mu           sync.Mutex
```
Comment (Benjamin2037) [inline br/pkg/lightning/backend/local/local.go:481]: Please also add some comment or give a self description name, instead of mu.

**Review by Benjamin2037 — APPROVED** [review]
for rest LGTM

File: br/pkg/lightning/backend/external/writer.go [hunk 5]
```diff
@@ -392,8 +409,15 @@ func (w *Writer) flushKVs(ctx context.Context, fromClose bool) (err error) {
 			zap.Any("rate", float64(savedBytes)/1024.0/1024.0/time.Since(ts).Seconds()))
 	}()
 
-	slices.SortFunc(w.writeBatch[:], func(i, j common.KvPair) int {
-		return bytes.Compare(i.Key, j.Key)
+	sorty.MaxGor = max(8, uint64(variable.GetDDLReorgWorkerCounter()))
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/writer.go:412]: maybe add a parameter to the builder to better reuse by other component

Comment (wjhuang2016) [reply br/pkg/lightning/backend/external/writer.go:412]: We can unify all the parameters in the later PR.

File: ddl/backfilling_operators.go [hunk 6]
```diff
@@ -148,21 +149,7 @@ func NewAddIndexIngestPipeline(
...
-	jobID int64,
-	subtaskID int64,
-	tbl table.PhysicalTable,
-	idxInfo *model.IndexInfo,
-	startKey, endKey kv.Key,
-	totalRowCount *atomic.Int64,
-	metricCounter prometheus.Counter,
-	onClose external.OnCloseFunc,
-) (*operator.AsyncPipeline, error) {
+func NewWriteIndexToExternalStoragePipeline(ctx *OperatorCtx, store kv.Storage, extStoreURI string, sessPool opSessPool, sessCtx sessionctx.Context, jobID, subtaskID int64, tbl table.PhysicalTable, idxInfo *model.IndexInfo, startKey, endKey kv.Key, totalRowCount *atomic.Int64, metricCounter prometheus.Counter, onClose external.OnCloseFunc, bcctx ingest.BackendCtx) (*operator.AsyncPipeline, error) {
```
Comment (D3Hunter) [inline ddl/backfilling_operators.go:152]: too long, prefer keep previous style

File: br/pkg/lightning/backend/external/writer.go [hunk 7]
```diff
@@ -360,6 +373,10 @@ func (w *Writer) flushKVs(ctx context.Context, fromClose bool) (err error) {
 	if len(w.writeBatch) == 0 {
 		return nil
 	}
+	if w.shareMu != nil {
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/writer.go:376]: we don't need this lock, if we're using 1 writer for each index

File: ddl/backfilling_operators.go [hunk 8]
```diff
@@ -453,27 +444,15 @@ type WriteExternalStoreOperator struct {
...
-	jobID int64,
-	subtaskID int64,
-	tbl table.PhysicalTable,
-	index table.Index,
-	store storage.ExternalStorage,
-	srcChunkPool chan *chunk.Chunk,
-	concurrency int,
-	onClose external.OnCloseFunc,
-) *WriteExternalStoreOperator {
+func NewWriteExternalStoreOperator(ctx *OperatorCtx, copCtx *CopContext, sessPool opSessPool, jobID int64, subtaskID int64, tbl table.PhysicalTable, index table.Index, store storage.ExternalStorage, srcChunkPool chan *chunk.Chunk, concurrency int, onClose external.OnCloseFunc, shareMu *sync.Mutex) *WriteExternalStoreOperator {
```
Comment (D3Hunter) [inline ddl/backfilling_operators.go:447]: doittoo

File: br/pkg/lightning/backend/external/writer.go (hunk 5 above)
Comment (Benjamin2037) [inline br/pkg/lightning/backend/external/writer.go:412]:
```suggestion
	sorty.MaxGor = min(8, uint64(variable.GetDDLReorgWorkerCounter()))
```

**Pull request merged**